bundles/org.openhab.binding.homeconnect/src/main/java/org/openhab/binding/homeconnect/internal/client/HomeConnectApiClient.java

```diff
@@ -0,0 +1,1134 @@
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.homeconnect.internal.client.exception.ApplianceOfflineException;
+import org.openhab.binding.homeconnect.internal.client.exception.AuthorizationException;
+import org.openhab.binding.homeconnect.internal.client.exception.CommunicationException;
+import org.openhab.binding.homeconnect.internal.client.model.ApiRequest;
+import org.openhab.binding.homeconnect.internal.client.model.AvailableProgram;
+import org.openhab.binding.homeconnect.internal.client.model.AvailableProgramOption;
+import org.openhab.binding.homeconnect.internal.client.model.Data;
+import org.openhab.binding.homeconnect.internal.client.model.HomeAppliance;
+import org.openhab.binding.homeconnect.internal.client.model.HomeConnectRequest;
+import org.openhab.binding.homeconnect.internal.client.model.HomeConnectResponse;
+import org.openhab.binding.homeconnect.internal.client.model.Option;
+import org.openhab.binding.homeconnect.internal.client.model.Program;
+import org.openhab.core.auth.client.oauth2.OAuthClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/**
+ * Client for Home Connect API.
+ *
+ * @author Jonas Brüstel - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class HomeConnectApiClient {
+    private static final String ACCEPT = "Accept";
+    private static final String CONTENT_TYPE = "Content-Type";
+    private static final String BSH_JSON_V1 = "application/vnd.bsh.sdk.v1+json";
+    private static final MediaType BSH_JSON_V1_MEDIA_TYPE = requireNonNull(MediaType.parse(BSH_JSON_V1));
+    private static final int REQUEST_READ_TIMEOUT = 30;
+    private static final int VALUE_TYPE_STRING = 0;
+    private static final int VALUE_TYPE_INT = 1;
+    private static final int VALUE_TYPE_BOOLEAN = 2;
+    private static final int COMMUNICATION_QUEUE_SIZE = 50;
+
+    private final Logger logger;
+    private final OkHttpClient client;
+    private final String apiUrl;
+    private final Map<String, List<AvailableProgramOption>> availableProgramOptionsCache;
+    private final OAuthClientService oAuthClientService;
+    private final Queue<ApiRequest> communicationQueue;
+    private final JsonParser jsonParser;
+
+    public HomeConnectApiClient(OAuthClientService oAuthClientService, boolean simulated,
+            @Nullable List<ApiRequest> apiRequestHistory) {
+        this.oAuthClientService = oAuthClientService;
+
+        availableProgramOptionsCache = new ConcurrentHashMap<>();
+        apiUrl = simulated ? API_SIMULATOR_BASE_URL : API_BASE_URL;
```

Comment:
I would prefer to have the URL handling at the very outset of the binding and use full URI/Strings throught the plumbing.
This would also make it possible for users to specify their own URLs for testing etc.

Reply:
The Home Connect ecosystem only uses two addresses. One for production and one for the simulated environment. There is no need to specify a custom URL.

Reply:
What about using a mock service for tests?
Or maybe someone reimplements the service as a self-hosted version.

Currently all low-level components have to contain the same conditional and know about all possible URLs.
When the URLs pulled farther up the stack (maybe even to the bridge settings itself)
it will be much easier to enable those usecases. It also simplifies to low-level components.

Reply:
> What about using a mock service for tests?

If you mock the Home Connect API (maybe in a unit test), you don't necessarily need to change the API URL. 


> Or maybe someone reimplements the service as a self-hosted version.

The Home Connect system is closed source. There are no open clients. At least I haven't heard of it yet. The official App is using a different protocol to communicate. 


> Currently all low-level components have to contain the same conditional and know about all possible URLs.

We are talking about two URLs. In the bridge configuration the user can decide whether to use the simulated or the productive environment. https://developer.home-connect.com/docs/general/endpoints_dataencoding


>When the URLs pulled farther up the stack (maybe even to the bridge settings itself)
it will be much easier to enable those usecases. It also simplifies to low-level components.

I think the bridge don't need to know details about the API. The knowledge of the API endpoint is not needed. The user shouldn't know about it. It would confuse the users if they can edit the URL in the bridge configuration.

Reply:
> > What about using a mock service for tests?
> If you mock the Home Connect API (maybe in a unit test), you don't necessarily need to change the API URL.
The imagination was to use something like wiremock to mock the API on the HTTP level.

> > Or maybe someone reimplements the service as a self-hosted version.
> The Home Connect system is closed source. There are no open clients. At least I haven't heard of it yet. The official App is using a different protocol to communicate.
This is more about an open API server.
Homekit is also closed source, and yet OpenHab implements it's interface.

> > When the URLs pulled farther up the stack (maybe even to the bridge settings itself)
> > it will be much easier to enable those usecases. It also simplifies to low-level components. 
> I think the bridge don't need to know details about the API. The knowledge of the API endpoint is not needed. The user shouldn't know about it. It would confuse the users if they can edit the URL in the bridge configuration.
It would be the default value of an advanced setting, not visible to users.
No component actually has to know about the distinction between prod and sandbox.

If you want, leave it as is.
It's just a pet-peeve of mine when enumerations of configurations are passed around instead of the raw configuration value.
It complicates the code and reduces the functionality without any upsides.